### PR TITLE
Introduce Cuttlefish

### PR DESCRIPTION
**Pull request by kuenishi**
This is still work in progress (testing is remaining), but just sending it loud as a pull request. I'd be more than happy if somebody take this. CC: https://github.com/basho/stanchion/pull/88

This pull request is now blocking other works, so
1. Review Stanchion config
2. Make an old/new configuration correspondence table [draft](https://github.com/basho/riak_cs/wiki/Configuration-Mapping-Table-between-1.5-and-2.0-%5BRFC%5D)
3. Review by CSE and update configuration

will follow this pull request.

### CODE REVIEW

**Comment by ksauzz** [comment]
- Updated rt runner https://github.com/basho/riak_test/pull/730
- Rebased the branch onto the current develop.

File: rel/files/riak_cs.schema [hunk 1]
```diff
@@ -0,0 +1,513 @@
...
+  {commented, "$(platform_etc_dir)/key.pem"}
+]}.
+
+%% @doc Enable this to allow the creation of an admin user when
+%% setting up a system. It is recommended to only enable this
+%% temporarily unless your use-case specifically dictates letting
+%% anonymous users to create accounts.
+{mapping, "anonymous_user_creation", "riak_cs.anonymous_user_creation", [
+  {default, false},
+  {datatype, flag}
```
Comment (kuenishi) [inline rel/files/riak_cs.schema:60]: isn't this should be `{flag, true, false}` ? [Flag description sounds like that](https://github.com/basho/cuttlefish/wiki/Datatypes#flag) but will test later.

Comment (kuenishi) [inline rel/files/riak_cs.schema:60]: If we move on to `on/off` vocabulary, default should be `off`.

File: rel/files/riak_cs.schema [hunk 2]
```diff
@@ -0,0 +1,513 @@
...
+]}.
+
+%% @doc Overflow pool size of secondary connection pool which is used
+%% strictly for requests to list the contents.
+{mapping, "pool.list.overflow", "riak_cs.connection_pools", [
+  {default, 0},
+  {datatype, integer}
+]}.
+
+{translation,
```
Comment (kuenishi) [inline rel/files/riak_cs.schema:129]: Nice

File: rel/files/riak_cs.schema [hunk 3]
```diff
@@ -0,0 +1,513 @@
...
+ end
+}.
+
+%% @doc Switch whether Riak CS trusts 'X-Forwarded-For' header.
+%% If your load balancer adds 'X-Forwarded-For' header
+%% and it is reliable (able to gaurantee it is not added
+%% by malicious user), turn this true. Otherwise, by
+%% default, Riak CS takes source IP address as an input.
+{mapping, "trust_x_forwarded_for", "riak_cs.trust_x_forwarded_for", [
+  {default, off},
```
Comment (kuenishi) [inline rel/files/riak_cs.schema:147]: Same as `anonymous_user_creation` argument.

File: rel/files/riak_cs.schema [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+{mapping, "gc.retry_interval", "riak_cs.gc_retry_interval", [
+  {default, "6h"},
+  {datatype, {duration, s}}
+]}.
+
+%% @doc Set this to false if you're running
+%% Riak nodes prior to version 1.4.0.
+{mapping, "gc.paginated_indexes", "riak_cs.gc_paginated_indexes", [
+  {default, on},
+  {datatype, flag}
```
Comment (kuenishi) [inline rel/files/riak_cs.schema:184]: It's good time to make this hidden.

File: rel/files/riak_cs.schema [hunk 5]
```diff
@@ -0,0 +1,513 @@
...
+{mapping, "auth_module", "riak_cs.auth_module", [
+  {default, riak_cs_s3_auth},
+  {datatype, atom}
+]}.
+
+%% == Bucket (List Objects) ==
+
+%% @doc Set this to false if you're running
+%% Riak nodes prior to version 1.4.0.
+{mapping, "fold_objects_for_list_keys", "riak_cs.fold_objects_for_list_keys", [
```
Comment (kuenishi) [inline rel/files/riak_cs.schema:307]: Same as `gc_pagenated_indexes` discussion.

**Comment by kuenishi** [comment]
We'd better update `src/riak_cs.app.src` and `rel/vars.config` as we changed `cs_ip` and `cs_port` to `listener` as a tuple, to avoid default confusion.

``` erlang
(riak-cs@127.0.0.1)2> [io:format("~p~n", [C])||C<-application:get_all_env(riak_cs)].
{auth_module,riak_cs_s3_auth}
{anonymous_user_creation,false}
{access_log_flush_size,1000000}
{access_archiver_max_backlog,2}
{stanchion_host,{"127.0.0.1",8085}}
{cs_root_host,"s3.amazonaws.com"}
{platform_log_dir,"./log"}
{gc_paginated_indexes,true}
{access_log_flush_factor,1}
{admin_key,"admin-key"}
{cs_port,8080}
{storage_archive_period,86400}
{cs_version,10300}
{access_archiver_max_workers,2}
{fold_objects_for_list_keys,true}
{admin_secret,"admin-secret"}
{stanchion_ssl,false}
{cs_ip,"127.0.0.1"}
{riak_host,{"127.0.0.1",8087}}
{gc_retry_interval,21600}
{trust_x_forwarded_for,false}
{dtrace_support,false}
{gc_interval,900}
{riakc_pool,{128,0}}
{connection_pools,[{request_pool,{128,0}},{bucket_list_pool,{5,0}}]}
{rewrite_module,riak_cs_s3_rewrite}
{leeway_seconds,86400}
{access_archive_period,3600}
{usage_request_limit,744}
{included_applications,[]}
{listener,{"127.0.0.1",8080}}
{put_fsm_buffer_size_max,10485760}
```

File: src/riak_cs_config.erl [hunk 6]
```diff
@@ -321,15 +322,10 @@ set_user_buckets_prune_time(_PruneTime) ->
...
-    Host = case application:get_env(riak_cs, riak_ip) of
-               {ok, Host0} -> Host0;
-               undefined -> "127.0.0.1"
-           end,
-    Port = case application:get_env(riak_cs, riak_pb_port) of
-               {ok, Port0 } -> Port0;
-               undefined -> 8087
-           end,
-    {Host, Port}.
+    case application:get_env(riak_cs, riak_host) of
```
Comment (kuenishi) [inline src/riak_cs_config.erl:332]:
``` erlang
get_env(riak_cs, riak_host, {"127.0.0.1", 8087}).
```

might be simpler.

File: riak_test/tests/storage_stats_test.erl [hunk 7]
```diff
@@ -233,8 +236,7 @@ calc_storage_stats() ->
     lager:info("riak-cs-storage batch result: ~s", [Res]),
     ExpectRegexp = "Batch storage calculation started.\n$",
     ?assertMatch({match, _}, re:run(Res, ExpectRegexp)),
-    %% TODO ensure storage calculation finished.
-    timer:sleep(1000),
+    true = rt:expect_in_log(CSNode, "Finished storage calculation"),
```
Comment (kuenishi) [inline riak_test/tests/storage_stats_test.erl:239]: Cool

File: riak_test/src/rtcs.erl [hunk 8]
```diff
@@ -885,55 +935,25 @@ error_child_element_verifier(Code, Message, Resource) ->
...
+%% @doc update config file from `From' to `To' version.
 migrate_cs(From, To, N, AdminCreds) ->
     migrate(From, To, N, AdminCreds, cs).
 
 migrate(From, To, N, AdminCreds, Who) when
       (From =:= current andalso To =:= previous)
       orelse ( From =:= previous andalso To =:= current) ->
-    Config = read_config(From, N, Who),
     Prefix = get_rt_config(Who, To),
+    Config = get_config(Who, To),
```
Comment (kuenishi) [inline riak_test/src/rtcs.erl]: This is for multibag flavor. mb test dopes mb configuration.. so `read_config` salvages those configurations. There might be other special configuration change involved. This `read_config` is for them.

File: rel/files/riak_cs.schema (hunk 1 above)
Comment (kuenishi) [inline rel/files/riak_cs.schema:60]: oh, true/false is working. Cool. I know little about cuttlefish.

**Comment by kuenishi** [comment]
Changes are mostly good! Nice work. Few things still left, my comment on `rtcs:read_config` and multibag flavour r_t.

**Comment by seancribbs** [comment]
Please be sure that a CSE has a chance to review the configuration before merging.

**Comment by kuenishi** [comment]
@seancribbs With this get merged, config items won't even be fixed. This is the first phase of cuttlefish work as this is rather a big change, blocking other works to avoid merge conflict confusion. I'll definitely create a correspondence table and sample .conf files with Stanchion involved, to get reviewed by CSEs.
That table will also be reusable for documentation.

**Comment by kuenishi** [comment]
With `{flavor, {multibag, disjoint}},`

```
Test Results:
user_test-cs_multi_backend                  : pass
upgrade_downgrade_test-cs_multi_backend     : fail
too_large_entity_test-cs_multi_backend      : pass
storage_stats_test-cs_multi_backend         : pass
stats_test-cs_multi_backend                 : pass
stanchion_switch_test-cs_multi_backend      : pass
sibling_benchmark-cs_multi_backend          : pass
repl_v3_test-cs_multi_backend               : pass
regression_tests_2-cs_multi_backend         : pass
regression_tests-cs_multi_backend           : pass
put_copy_test-cs_multi_backend              : pass
object_get_test-cs_multi_backend            : pass
object_get_conditional_test-cs_multi_backend: pass
multibag_transition_test-cs_multi_backend   : pass
multibag_disjoint_test-cs_multi_backend     : pass
mp_upload_test-cs_multi_backend             : pass
list_objects_v2_test-cs_multi_backend       : pass
list_objects_test-cs_multi_backend          : pass
legacy_s3_rewrite_test-cs_multi_backend     : pass
gc_tests-cs_multi_backend                   : pass
external_client_tests-cs_multi_backend      : pass
cs743_regression_test-cs_multi_backend      : pass
buckets_test-cs_multi_backend               : pass
auth_bypass_test-cs_multi_backend           : pass
access_stats_test-cs_multi_backend          : pass
---------------------------------------------
1 Tests Failed
24 Tests Passed
That's 96.0% for those keeping score
```

File: rel/files/riak_cs.schema (hunk 1 above)
Comment (ksauzz) [inline rel/files/riak_cs.schema:60]: `on/off` looks proper values. I should change `false` to `off`.

File: rel/files/riak_cs.schema (hunk 4 above)
Comment (ksauzz) [inline rel/files/riak_cs.schema:184]: Sounds reasonable to me. I'll do that.

**Comment by ksauzz** [comment]
hmmm, it works for me with `{flavor, {multibag, disjoint}},`. How did upgrade_downgrade_test fail ?
Could you show me the error message ?

```
Test Results:
upgrade_downgrade_test-cs_multi_backend: pass
---------------------------------------------
0 Tests Failed
1 Tests Passed
```

**Comment by kuenishi** [comment]
Still failing for disjoint.
```
17:18:19.419 [debug] updating /home/kuenishi/rt/riak_cs/current/dev/dev1/etc/app.config
17:18:19.419 [error] Error in process <0.135.0> on node 'riak_test@127.0.0.1' with exit value: {{badmatch,{error,enoent}},[{rtcs,update_app_config,2,[{file,"r
iak_test/src/rtcs.erl"},{line,738}]},{upgrade_downgrade_test,'-confirm/0-lc$^0/1-0-',3,[{file,"riak_test/tests/upgrade_downgrade_test.erl"},{line,53}]},{upgra
de_downgrade_test... 


17:18:19.419 [warning] upgrade_downgrade_test failed: {{badmatch,{error,enoent}},[{rtcs,update_app_config,2,[{file,"riak_test/src/rtcs.erl"},{line,738}]},{upg
rade_downgrade_test,'-confirm/0-lc$^0/1-0-',3,[{file,"riak_test/tests/upgrade_downgrade_test.erl"},{line,53}]},{upgrade_downgrade_test,confirm,0,[{file,"riak_
test/tests/upgrade_downgrade_test.erl"},{line,47}]},{riak_test_runner,return_to_exit,3,[{file,"src/riak_test_runner.erl"},{line,159}]}]}
17:18:19.420 [error] 
================ upgrade_downgrade_test failure stack trace =====================
{{badmatch,{error,enoent}},
 [{rtcs,update_app_config,2,[{file,"riak_test/src/rtcs.erl"},{line,738}]},
  {upgrade_downgrade_test,'-confirm/0-lc$^0/1-0-',3,
                          [{file,"riak_test/tests/upgrade_downgrade_test.erl"},
                           {line,53}]},
  {upgrade_downgrade_test,confirm,0,
                          [{file,"riak_test/tests/upgrade_downgrade_test.erl"},
                           {line,47}]},
  {riak_test_runner,return_to_exit,3,
                    [{file,"src/riak_test_runner.erl"},{line,159}]}]}
=================================================================================
```

**Comment by kuenishi** [comment]
Failed for me...

```
17:24:09.075 [warning] upgrade_downgrade_test failed: {{aws_error,{http_error,404,[],<<"<?xml version=\"1.0\" encoding=\"UTF-8\"?><Error><Code>NoSuchKey</Code
><Message>The specified key does not exist.</Message><Resource>/riak-test-bucket-foobar/riak_test_key1</Resource><RequestId></RequestId></Error>">>}},[{erlclo
ud_s3,s3_request,9,[{file,"src/erlcloud_s3.erl"},{line,898}]},{erlcloud_s3,fetch_object,5,[{file,"src/erlcloud_s3.erl"},{line,456}]},{upgrade_downgrade_test,b
asic_get_test_case,4,[{file,"riak_test/tests/upgrade_downgrade_test.erl"},{line,145}]},{upgrade_downgrade_test,verify_all_data,2,[{file,"riak_test/tests/upgra
de_downgrade_test.erl"},{line,133}]},{upgrade_downgrade_test,confirm,0,[{file,"riak_test/tests/upgrade_downgrade_test.erl"},{line,62}]},{riak_test_runner,retu
rn_to_exit,3,[{file,"src/riak_test_runner.erl"},{line,159}]}]}
17:24:09.076 [error] 
================ upgrade_downgrade_test failure stack trace =====================
{{aws_error,{http_error,404,[],
                        <<"<?xml version=\"1.0\" encoding=\"UTF-8\"?><Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message><Resourc
e>/riak-test-bucket-foobar/riak_test_key1</Resource><RequestId></RequestId></Error>">>}},
 [{erlcloud_s3,s3_request,9,[{file,"src/erlcloud_s3.erl"},{line,898}]},
  {erlcloud_s3,fetch_object,5,[{file,"src/erlcloud_s3.erl"},{line,456}]},
  {upgrade_downgrade_test,basic_get_test_case,4,
                          [{file,"riak_test/tests/upgrade_downgrade_test.erl"},
                           {line,145}]},
  {upgrade_downgrade_test,verify_all_data,2,
                          [{file,"riak_test/tests/upgrade_downgrade_test.erl"},
                           {line,133}]},
  {upgrade_downgrade_test,confirm,0,
                          [{file,"riak_test/tests/upgrade_downgrade_test.erl"},
                           {line,62}]},
  {riak_test_runner,return_to_exit,3,
                    [{file,"src/riak_test_runner.erl"},{line,159}]}]}
=================================================================================

17:24:09.076 [notice] upgrade_downgrade_test Test Run Complete
17:24:09.145 [info] Stopping cover

Test Results:
upgrade_downgrade_test-cs_multi_backend: fail
---------------------------------------------
1 Tests Failed
0 Tests Passed
That's 0.0% for those keeping score
```

I did refresh the deps and re-run stagedevrel and compile-riak-test, and rtdev-current. I think I've all rebuilt my rt env.

**Comment by kuenishi** [comment]
This is the advanced.config right when test failed after upgrade. Multibag configurations have been lost and that's why test is not working.

``` erlang
[{lager,
     [{handlers,
          [{lager_file_backend,
               [{"./log/error.log",error,10485760,"$D0",5},
                {"./log/console.log",debug,10485760,"$D0",5}]}]}]},
 {riak_cs,
     [{admin_key,"VI62FXNELRYLR7S0ALUB"},
      {admin_secret,"seBRN1sXAtWMp4uQbZ9byUk8KovRpsQLCAEX0Q=="},
      {anonymous_user_creation,true},
      {block_get_max_retries,1},
      {connection_pools,[{request_pool,{8,0}},{bucket_list_pool,{2,0}}]},
      {cs_port,15018},
      {cs_version,10300},
      {proxy_get,enabled},
      {riak_host,{"127.0.0.1",10017}},
      {stanchion_host,{"127.0.0.1",9095}}]}].
```

**Comment by kuenishi** [comment]
So, my suggestion is still to replace `get_config/2` with `read_config/1` .

**Comment by kuenishi** [comment]
@borshop merge

**Comment by Basho-JIRA** [comment]
Yay!! 

**_[posted via JIRA by Kota Uenishi]**_

**Pull request merged**